src/internal/common/limiters/sliding_window.go

```diff
@@ -0,0 +1,197 @@
+	mu sync.Mutex
+	// stopTimer stops the recurring slide timer
+	stopTimer chan struct{}
+}
+
+func NewSlidingWindowLimiter(
+	windowSize, slideInterval time.Duration,
+	capacity int,
+) (Limiter, error) {
+	if err := validate(windowSize, slideInterval, capacity); err != nil {
+		return nil, err
+	}
+
+	ni := int(windowSize / slideInterval)
+
+	s := &slidingWindow{
+		windowSize:    windowSize,
+		slideInterval: slideInterval,
+		capacity:      capacity,
+		permits:       make(chan token, capacity),
+		numIntervals:  ni,
+		prev: fixedWindow{
+			count: make([]int, ni),
+		},
+		curr: fixedWindow{
+			count: make([]int, ni),
+		},
+		currentInterval: -1,
+		stopTimer:       make(chan struct{}),
+	}
+
+	s.initialize()
+
+	return s, nil
+}
+
+// Wait blocks a request until a token is available or the context is cancelled.
+// TODO(pandeyabs): Implement WaitN.
+func (s *slidingWindow) Wait(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-s.permits:
+		s.mu.Lock()
+		defer s.mu.Unlock()
+
+		s.curr.count[s.currentInterval]++
+	}
+
+	return nil
+}
+
+// Shutdown cleans up the slide goroutine. If shutdown is not called, the slide
+// goroutine will continue to run until the program exits.
+func (s *slidingWindow) Shutdown() {
+	select {
+	case s.stopTimer <- struct{}{}:
+	default:
+	}
+}
+
+// initialize starts the slide goroutine and prefills tokens to full capacity.
+func (s *slidingWindow) initialize() {
+	// Ok to not hold the mutex here since nothing else is running yet.
+	s.nextInterval()
+
+	// Start a goroutine which runs every slideInterval. This goroutine will
+	// continue to run until the program exits or until Shutdown is called.
+	go func() {
+		ticker := time.NewTicker(s.slideInterval)
+
+		for {
+			select {
+			case <-ticker.C:
+				s.slide()
+			case <-s.stopTimer:
+				ticker.Stop()
+				return
+			}
+		}
+	}()
+
+	// Prefill permits to allow tokens to be granted immediately
+	for i := int(0); i < s.capacity; i++ {
+		s.permits <- token{}
+	}
+}
+
+// nextInterval increments the current interval and slides the fixed
+// windows if needed. Should be called with the mutex held.
+func (s *slidingWindow) nextInterval() {
+	// Increment current interval
+	s.currentInterval = (s.currentInterval + 1) % s.numIntervals
+
+	// Slide the fixed windows if windowSize time has elapsed.
+	if s.currentInterval == 0 {
+		s.prev = s.curr
+		s.curr = fixedWindow{
+			count: make([]int, s.numIntervals),
+		}
+	}
+}
+
+// slide moves the window forward by one interval. It reclaims tokens from the
+// interval that we slid past and adds them back to available permits. If the
+// permits are already at capacity, excess tokens are discarded.
+func (s *slidingWindow) slide() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.nextInterval()
+
+	for i := int(0); i < s.prev.count[s.currentInterval]; i++ {
+		select {
+		case s.permits <- token{}:
+		default:
+			// Skip if permits are at capacity
+			return
+		}
```

Comment:
this feels a little odd to me. If we're never creating more tokens once the limiter's been initialized then how can the channel be at capacity if the channel itself can buffer as many tokens as we have total? It seems like either it's never going to pick the default case because the channel is buffered or if it does pick the default case without us knowing then we'll permanently lose some capacity in the rate limiter because we'll never add those tokens back

There's a quick example [here](https://play.golang.com/p/8RHMrQO36oi) that you could play around with if you're interested. Adjusting the buffer on the channel can cause the total number of sends on it to change which causes the final "count" printed at the end to change

Reply:
> If we're never creating more tokens once the limiter's been initialized then how can the channel be at capacity if the channel itself can buffer as many tokens as we have total?

Great catch! The default case will never execute here. Like you said, total tokens at any time = tokens buffered in the `permits` channel + tokens consumed in the sliding window. If the permits channel is at capacity, then the `s.prev.count[s.currentInterval]` can never have non-zero value.